core/src/session.rs

```diff
@@ -252,8 +252,20 @@ impl Session {
             if session.is_invalid() {
                 break;
             }
-            let last_ping = session.0.data.read().last_ping.unwrap_or_else(Instant::now);
-            if last_ping.elapsed() >= SESSION_TIMEOUT {
+
+            // It would be so much easier to use elapsed but elapsed could
+            // potentially panic is rare cases.
+            // See:
+            // https://doc.rust-lang.org/std/time/struct.Instant.html#monotonicity
```

Comment:
I guess this can only hurt us when running virtualised?
We should not care about any platforms below Rust tier 1.

Reply:
If it's this easy to care why not do it right?

Reply:
It also does not say that the correct behavior is even guaranteed on tier 1.

"On all platforms Instant will try to use an OS API that guarantees monotonic behavior if available, which is the case for all [tier 1](https://doc.rust-lang.org/rustc/platform-support.html) platforms. In practice such guarantees are – under rare circumstances – broken by hardware, virtualization or operating system bugs."

The operative word is "try" and as mentioned it can panic even on tier 1.

"To work around these bugs and platforms not offering monotonic clocks [duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since), [elapsed](https://doc.rust-lang.org/std/time/struct.Instant.html#method.elapsed) and [sub](https://doc.rust-lang.org/std/time/struct.Instant.html#method.sub) saturate to zero. In older Rust versions this lead to a panic instead. [checked_duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_duration_since) can be used to detect and handle situations where monotonicity is violated, or Instants are subtracted in the wrong order."

The way I'm doing it is the safest and most correct way as far as the docs are concerned.

I read all of that to say that if you care at all about not panicking just don't ever use elapsed.

Reply:
If you also look at the docs for elapsed the behavior is inconsistent across different Rust versions and subject to change:

"[Panics](https://doc.rust-lang.org/std/time/struct.Instant.html#panics-1)
Previous rust versions panicked when the current time was earlier than self. Currently this method returns a Duration of zero in that case. Future versions may reintroduce the panic. See [Monotonicity](https://doc.rust-lang.org/std/time/struct.Instant.html#monotonicity)."

That makes it a ticking time bomb in the code.

Reply:
If the current behavior is correct we should probably keep track of other clients so we can be like "Yep, we already responded to that Hello, we don't need to notify over and over again forever."

Reply:
OK. Is there even a PR or issue open for this in Rust? Doesn't seem right to me that such a core function can panic without having a panic-free version of it.

Reply:
I don't know? I just came upon it when reworking the "should notify" code in player and was pretty shocked myself that something that you would think would be very basic has so many gotchas. The fact that the docs mention it would indicate to me that it's probably at least been brought up? I don't know what the current state of things are though as far as if anyone is still worried about it? I just plan on just never using elapsed from now on.